invenio_files_rest/models.py

```diff
@@ -408,27 +429,30 @@ def validate_storage_class(self, key, default_storage_class):
             raise ValueError('Invalid storage class.')
         return default_storage_class
 
-    @ensure_not_deleted()
     def snapshot(self, lock=False):
         """Create a snapshot of latest objects in bucket.
 
         :param lock: Create the new bucket in a locked state.
         :returns: Newly created bucket with the snapshot.
         """
         with db.session.begin_nested():
-            b = Bucket(
+            bucket = Bucket(
                 default_location=self.default_location,
                 default_storage_class=self.default_storage_class,
                 quota_size=self.quota_size,
             )
-            db.session.add(b)
+            db.session.add(bucket)
+        return self.merge(lock=lock, bucket=bucket)
 
-        for o in ObjectVersion.get_by_bucket(self):
-            o.copy(bucket=b)
+    @ensure_not_deleted()
+    def merge(self, bucket, lock=False):
+        """Sync self to the destination bucket."""
+        for o in ObjectVersion.get_heads_by_bucket(self):
+            o.merge(bucket=bucket, key=o.key)
 
-        b.locked = True if lock else self.locked
+        bucket.locked = lock or self.locked
```

Comment:
I don't know which should be the correct way of using the `lock` parameter, if I set it to `False` I think that the returning bucket should be `bucket.locked = False`.
Otherwise you should always do
```python
new_bucket = bucket.snapshot()
new_bucket.locked = False
```
@switowski WDYT?